miden/src/lib.rs

```diff
@@ -1,84 +1,70 @@
-use air::{ProcessorAir, PublicInputs, TraceMetadata, TraceState, MAX_OUTPUTS, MIN_TRACE_LENGTH};
-use core::convert::TryInto;
+use air::{ProcessorAir, PublicInputs};
+use processor::{ExecutionError, ExecutionTrace};
+use prover::Prover;
+use vm_core::{Felt, StarkField, MIN_STACK_DEPTH};
+
 #[cfg(feature = "std")]
 use log::debug;
-use prover::{Prover, ProverError, Serializable, Trace, TraceTable};
+#[cfg(feature = "std")]
+use prover::Trace;
 #[cfg(feature = "std")]
 use std::time::Instant;
 
-#[cfg(test)]
-mod tests;
-
 // EXPORTS
 // ================================================================================================
 
 pub use air::{FieldExtension, HashFunction, ProofOptions};
-pub use assembly;
-pub use processor::{BaseElement, FieldElement, Program, ProgramInputs, StarkField};
+pub use assembly::{Assembler, AssemblyError};
 pub use prover::StarkProof;
-pub use verifier::{verify, VerifierError};
+pub use verifier::{verify, VerificationError};
+pub use vm_core::{program::Script, ProgramInputs};
 
 // EXECUTOR
 // ================================================================================================
 
-/// Executes the specified `program` and returns the result together with a STARK-based proof of execution.
+/// Executes the specified `program` and returns the result together with a STARK-based proof of
+/// the program's execution.
 ///
-/// * `inputs` specifies the initial stack state and provides secret input tapes;
-/// * `num_outputs` specifies the number of elements from the top of the stack to be returned;
+/// * `inputs` specifies the initial state of the stack as well as non-deterministic (secret)
+///   inputs for the VM.
+/// * `num_stack_outputs` specifies the number of elements from the top of the stack to be
+///   returned.
+/// * `options` defines parameters for STARK proof generation.
+///
+/// # Errors
+/// Returns an error if program execution or STARK proof generation fails for any reason.
 pub fn execute(
-    program: &Program,
+    program: &Script,
```

Comment:
I'm interested in the motivation for this name change, because it seems less clear to me, especially since we're calling it ```program```, and we have ```ProgramInputs```. Also, I've been thinking of scripts the way it's defined in the Assembly specs, which excludes procedures, but that would not be the case here. I.e. procedures could be included in this Script, since they are included in compile_script (which this comment probably also applies to). That means we have 2 different types of scripts in our conceptual framework.

Reply:
The idea was that we'd have two different types of programs `Scripts` and `Modules`. Scripts are basically just code while modules would be code + data. But then I decided to delay working on modules to v0.3 (we need to have storage for modules to work) - and so at this point scripts are basically equivalent to the old `Program`.

I think we'll need to come back to this when we start working on v0.3. For example, the word "module" is now overloaded since we have modules in the standard library, and maybe we should use something else for the original modules I had in mind.

Reply:
Thanks for the explanation. Revisiting this later sounds good